docker/docker_client.go

```diff
@@ -447,7 +447,37 @@ func (c *dockerClient) makeRequest(ctx context.Context, method, path string, hea
 	}
 
 	url := fmt.Sprintf("%s://%s%s", c.scheme, c.registry, path)
-	return c.makeRequestToResolvedURL(ctx, method, url, headers, stream, -1, auth, extraScope)
+	resp, err := c.makeRequestToResolvedURL(ctx, method, url, headers, stream, -1, auth, extraScope)
+
+	// As by default the challenges from the /v2/ "ping" response were used for fetching the
+	// auth token, we might be getting a error back indicating that the auth scope was insufficient
+	// check for that and update the client challenges to retry with requesting a new token
+	if retry, challenges := needsRetryWithUpdatedChallenges(err, resp); retry {
+		logrus.Debug("Detected insufficent_scope error, retrying request with updated challenges")
+		c.challenges = challenges
```

Comment:
`dockerClient`, once initialized, can be used concurrently (`HasThreadSafe…`), so overwriting a property which is intentionally in the “… and never change afterwards” section of `dockerClient` is not safe. (This is also why `extraScope` is a parameter and not a property.)

Reply:
Ah, ok. Must have overlooked that :(. I had initially thought to put the updated scope into the `extraScope` argument, but as that would have required to rework the `authScope` struct a bit (as the current code always assumes the resourcetype part of it to be `repository` I went of the above approach in my first try. I guess I need to rethink that again :).

 What do you think, does it make sense to pass the updated scope via the `extraScope` argument?